File: stack.java
import java.util.*;

public class stack {
    static int precedence(String op) {
        if (op.equals("+") || op.equals("-")) return 1;
        if (op.equals("*") || op.equals("/")) return 2;
        return 0;
    }

    public static void main(String[] args) {
        Scanner sc = new Scanner(System.in);
        String loop;

        do {
            System.out.println("\n========================================");
            System.out.print("Masukkan ekspresi aritmatika: ");
            String input = sc.nextLine();

            // Tokenizing menggunakan regex agar support input tanpa spasi
            String[] tokens = input.replaceAll("([+\\-*/()])", " $1 ").trim().split("\\s+");

            Stack<String> stack = new Stack<>();
            List<String> postfix = new ArrayList<>();

            System.out.println("\n--- Tahap 1: Infix -> Postfix ---");
            for (String t : tokens) {
                if (t.matches("\\d+")) {
                    postfix.add(t);
                } else if (t.equals("(")) {
                    stack.push(t);
                } else if (t.equals(")")) {
                    while (!stack.isEmpty() && !stack.peek().equals("(")) postfix.add(stack.pop());
                    if (!stack.isEmpty()) stack.pop();
                } else {
                    while (!stack.isEmpty() && precedence(stack.peek()) >= precedence(t)) {
                        postfix.add(stack.pop());
                    }
                    stack.push(t);
                }
                System.out.println("Token: " + t + "\t| Stack: " + stack + " | Postfix: " + postfix);
            }

            // Cleanup sisa stack
            while (!stack.isEmpty()) {
                postfix.add(stack.pop());
                System.out.println("Cleanup\t| Stack: " + stack + " | Postfix: " + postfix);
            }

            System.out.println("\n--- Tahap 2: Evaluasi ---");
            Stack<Double> eval = new Stack<>();
            try {
                for (String t : postfix) {
                    if (t.matches("\\d+")) {
                        eval.push(Double.parseDouble(t));
                    } else {
                        double b = eval.pop();
                        double a = eval.pop();
                        double res = 0;
                        switch (t) {
                            case "+": res = a + b; break;
                            case "-": res = a - b; break;
                            case "*": res = a * b; break;
                            case "/": res = a / b; break;
                        }
                        eval.push(res);
                        System.out.println("Operasi " + a + t + b + " = " + res + " | Stack: " + eval);
                    }
                }
                System.out.println("\nHASIL AKHIR: " + eval.peek());
            } catch (Exception e) {
                System.out.println("Input tidak valid!");
            }

            System.out.print("\nHitung lagi? (y/n): ");
            loop = sc.nextLine().toLowerCase();
        } while (loop.equals("y"));

        System.out.println("Selesai.");
        sc.close();
    }
}

File: stack.py
import re

def get_precedence(op):
    # Menentukan prioritas operator (Higher = Stronger)
    return {'+': 1, '-': 1, '*': 2, '/': 2}.get(op, 0)

def main():
    while True:
        print("\n" + "="*50)
        infix = input("Masukkan ekspresi aritmatika (atau ketik 'exit' untuk keluar): ")
        
        if infix.lower() == 'exit':
            break
            
        # Regex untuk memisahkan angka (multi-digit) dan operator
        tokens = re.findall(r'\d+|[+\-*/()]', infix)
        
        stack = []
        postfix = []

        print(f"\n--- TAHAP 1: Konversi Infix ke Postfix (Shunting-yard) ---")
        print(f"{'Token':<10} | {'Stack':<20} | {'Postfix List'}")
        print("-" * 55)

        for token in tokens:
            if token.isdigit():
                postfix.append(token)
            elif token == '(':
                stack.append(token)
            elif token == ')':
                while stack and stack[-1] != '(':
                    postfix.append(stack.pop())
                    print(f"{' ':10} | {str(stack):20} | {' '.join(postfix)} (Pop karena ')')")
                if stack: stack.pop() # Buang '('
            else:
                while stack and stack[-1] != '(' and get_precedence(stack[-1]) >= get_precedence(token):
                    op = stack.pop()
                    postfix.append(op)
                    print(f"{' ':10} | {str(stack):20} | {' '.join(postfix)} (Pop '{op}' prioritas)")
                stack.append(token)
            
            print(f"{token:<10} | {str(stack):20} | {' '.join(postfix)}")

        # Proses Cleanup: Menguras sisa stack
        while stack:
            op = stack.pop()
            postfix.append(op)
            print(f"{'Cleanup':<10} | {str(stack):20} | {' '.join(postfix)} (Pop sisa '{op}')")
        
        print(f"\n>> Postfix Final: {' '.join(postfix)}")

        # TAHAP 2: EVALUASI
        eval_stack = []
        print(f"\n--- TAHAP 2: Evaluasi Perhitungan (Stack) ---")
        try:
            for token in postfix:
                if token.isdigit():
                    eval_stack.append(float(token))
                    print(f"Push {token:4} | Stack: {eval_stack}")
                else:
                    v2 = eval_stack.pop() # Operan kanan
                    v1 = eval_stack.pop() # Operan kiri
                    
                    if token == '+': res = v1 + v2
                    elif token == '-': res = v1 - v2
                    elif token == '*': res = v1 * v2
                    elif token == '/': res = v1 / v2
                    
                    eval_stack.append(res)
                    print(f"Hitung {v1} {token} {v2} | Hasil: {res} | Stack: {eval_stack}")
            
            print(f"\nHASIL AKHIR: {eval_stack[0]}")
        except Exception:
            print("Error: Ekspresi tidak valid!")

        if input("\nHitung lagi? (y/n): ").lower() != 'y':
            print("Program Berhenti.")
            break

if __name__ == "__main__":
    main()